tun/tun_illumos.go

```diff
@@ -0,0 +1,271 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2019 Joyent, Inc.
+ */
+
+// +build illumos,amd64
+
+package tun
+
+import (
+	"fmt"
+
+	"golang.org/x/sys/unix"
+	"os"
+)
+
+// NativeTun is the OS-specific implementation of a tun interface.
+type NativeTun struct {
+	tunFile *os.File /* TUN device file */
+	ip_fd   int      /* IP device fd */
+	name    string   /* Interface name */
+	mtu     int
+
+	events chan Event
+	errors chan error
+}
+
+// Name implements the Device interface
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+// Flush implements the Device interface
+func (tun *NativeTun) Flush() error {
+	// TODO: can flushing be implemented by buffering and using sendmmsg?
+	return nil
+}
+
+// Read implements the Device interface
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(int(tun.tunFile.Fd()), nil, buf[offset:])
+		if err != nil {
+			return 0, err
+		}
+		return len(read), nil
+	}
+}
+
+// Write implements the Device interface
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	if err := unix.Putmsg(int(tun.tunFile.Fd()), nil, buf[offset:], 0); err != nil {
+		return 0, err
+	}
+	return len(buf), nil
+}
+
+// File implements the Device interface
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+// MTU implements the Device interface
+func (tun *NativeTun) MTU() (int, error) {
+	return tun.mtu, nil
+}
+
+// Events implements the Device interface
+func (tun *NativeTun) Events() chan Event {
+	return tun.events
+}
+
+// Close implements the Device interface
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
+
+		var l unix.Lifreq
+		if err := l.SetName(tun.name); err != nil {
+			return err
+		}
+
+		// illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint. Hand-cast things to help
+		// the compiler figure it out.
+		reqnum := int(unix.SIOCGLIFMUXID)
+		if err := unix.IoctlLifreq(tun.ip_fd, uint(reqnum), &l); err != nil {
+			return err
+		}
+
+		id := l.GetLifruInt()
+
+		if err := unix.IoctlSetInt(tun.ip_fd, unix.I_PUNLINK, id); err != nil {
+			return err
+		}
+
+		unix.Close(tun.ip_fd)
+		tun.ip_fd = -1
+	}
+
+	if tun.tunFile != nil {
+		tun.tunFile.Close()
+		tun.tunFile = nil
```

Comment:
If you're going to use `SyscallConn`, as suggested above, maybe don't set this to nil here.